### PR TITLE
add taxa freq loader

### PR DESCRIPTION
**Pull request by Tianhao-Gu**

### CODE REVIEW

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                   os.path.join(self.caller_file_dir, 'ar53_taxonomy_r207.tsv'),
+                   str(release_version),
+                   '--output', result_file,
+                   '--kbase_collection', kbase_collections]
+
+        p = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
+        stdout, stderr = p.communicate()
+        print(str(stdout), str(stderr))
+        with open(result_file, 'r') as f:
+            data = json.load(f)
```
Comment (MrCreosote) [inline test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py:69]: I think these lines are the same as the prior test, extract into a helper method if so

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        expected_docs_length = 5420
+        assert len(data) == expected_docs_length
+
+        first_doc = data[0]
+        expected_keys = {'_key', 'collection', 'load_version', 'rank', 'name', 'count'}
+        self.assertCountEqual(first_doc.keys(), expected_keys)
```
Comment (MrCreosote) [inline test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py:76]: Same question as prior test

File: src/loaders/gtdb/gtdb_taxa_freq_loader.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+
+def _convert_to_json(docs, outfile):
+    json_object = json.dumps(docs, indent=4)
+    outfile.write(json_object)
+    outfile.close()
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Prepare GTDB taxa frequency data in JSON format for arango import.')
```
Comment (MrCreosote) [inline src/loaders/gtdb/gtdb_taxa_freq_loader.py:83]: Can you stick PROTOTYPE somewhere in there, I want to make sure that before we release anything we run through the code again and make sure it's all g2g and has good tests

**Review by MrCreosote — CHANGES_REQUESTED** [review]
Generally looks good

File: src/loaders/gtdb/gtdb_taxa_freq_loader.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+"""
+Prepare GTDB taxa frequency data in JSON format for arango import.
+
+usage: gtdb_taxa_freq_loader.py [-h]
+                              [--kbase_collection KBASE_COLLECTION]
+                              [-o OUTPUT]
+                              load_files
+                              [load_files ...]
+                              release_version
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/gtdb/gtdb_taxa_freq_loader.py:14]: 👍

File: src/loaders/gtdb/gtdb_taxa_freq_loader.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+
+def _get_lineage(linstr):
+    ln = linstr.split(";")
+    ret = []
+    for lin in ln:
+        taxa_abbrev, taxa_name = lin.split("__")
+        ret.append({"abbrev": taxa_abbrev, "name": taxa_name})
+    if ret[-1]["abbrev"] != "s":
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/gtdb/gtdb_taxa_freq_loader.py:40]: 👍

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import json
+import os
+import shutil
+import subprocess
+import unittest
+import uuid
+from pathlib import Path
+
+
+class TaxaFrequencyTest(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py:11]: 👍

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        print(str(stdout), str(stderr))
+        with open(result_file, 'r') as f:
+            data = json.load(f)
+
+        expected_docs_length = 5420
+        assert len(data) == expected_docs_length
+
+        first_doc = data[0]
+        expected_keys = {'_key', 'collection', 'load_version', 'rank', 'name', 'count'}
+        self.assertCountEqual(first_doc.keys(), expected_keys)
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py:47]: 👍

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py (hunk 1 above)
Comment (Tianhao-Gu) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py:69]: 👍

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py (hunk 2 above)
Comment (Tianhao-Gu) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py:76]: 👍

File: src/loaders/gtdb/gtdb_taxa_freq_loader.py (hunk 3 above)
Comment (Tianhao-Gu) [reply src/loaders/gtdb/gtdb_taxa_freq_loader.py:83]: 👍

File: src/loaders/gtdb/gtdb_taxa_freq_loader.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                "name": name,
+                "count": nodes[rank][name]
+            }
+            docs.append(doc)
+
+    return docs
+
+
+def _convert_to_json(docs, outfile):
+    json_object = json.dumps(docs, indent=4)
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/gtdb/gtdb_taxa_freq_loader.py:77]: 👍

File: src/loaders/gtdb/gtdb_taxa_freq_loader.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/gtdb/gtdb_taxa_freq_loader.py]: :+1:

File: src/loaders/gtdb/gtdb_taxa_freq_loader.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/gtdb/gtdb_taxa_freq_loader.py]: :+1:

File: src/loaders/gtdb/gtdb_taxa_freq_loader.py (hunk 3 above)
Comment (MrCreosote) [reply src/loaders/gtdb/gtdb_taxa_freq_loader.py:86]: Can you add it to the help message too please?

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+import shutil
+import subprocess
+import uuid
+from pathlib import Path
+
+import jsonlines
+import pytest
+
+
+@pytest.fixture
```
Comment (MrCreosote) [inline test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py:12]: I think you want to do
```suggestion
@pytest.fixture(scope="module")
```
so this only runs once total, not once per test

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py]: :+1:

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    print('starting GTDB taxa frequency test')
+    tmp_dir = 'result_{}'.format(str(uuid.uuid4()))
+    os.makedirs(tmp_dir, exist_ok=True)
+
+    caller_frame = inspect.stack()[0]
+    caller_filename_full = caller_frame.filename
+    caller_file_dir = os.path.dirname(caller_filename_full)
+
+    project_dir = Path(caller_file_dir).resolve().parents[3]
+    script_file = '{}/src/loaders/gtdb/gtdb_taxa_freq_loader.py'.format(project_dir)
```
Comment (MrCreosote) [inline test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py:23]:
```suggestion
    script_file = f'{project_dir}/src/loaders/gtdb/gtdb_taxa_freq_loader.py'
```

File: src/loaders/gtdb/gtdb_taxa_freq_loader.py (hunk 3 above)
Comment (Tianhao-Gu) [reply src/loaders/gtdb/gtdb_taxa_freq_loader.py]: yup. added.

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py (hunk 9 above)
Comment (Tianhao-Gu) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py]: 👍

**Review by MrCreosote — COMMENTED** [review]
One other minor change

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py]: :+1:

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py (hunk 1 above)
Comment (MrCreosote) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py]: :+1:

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py (hunk 2 above)
Comment (MrCreosote) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py]: :+1:

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py (hunk 10 above)
Comment (Tianhao-Gu) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py]: 👍 sure.

File: src/loaders/gtdb/gtdb_taxa_freq_loader.py (hunk 3 above)
Comment (MrCreosote) [reply src/loaders/gtdb/gtdb_taxa_freq_loader.py]: :+1:

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py (hunk 9 above)
Comment (MrCreosote) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py]: :+1:

File: test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py (hunk 10 above)
Comment (MrCreosote) [reply test/src/loaders/gtdb/gtdb_taxa_freq_loader_test.py]: :+1:

**Review by MrCreosote — APPROVED** [review]
LGTM

**Pull request merged**